staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/client.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vault implements envelop encryption provider based on Vault KMS
+package vault
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"path"
+	"reflect"
+
+	vaultapi "github.com/hashicorp/vault/api"
+)
+
+const defaultTransitPath = "transit"
+const defaultAuthPath = "auth"
+
+// Handle all communication with Vault server.
+type clientWrapper struct {
+	client      *vaultapi.Client
+	encryptPath string
+	decryptPath string
+	authPath    string
+}
+
+// Initialize a client wrapper for vault kms provider.
+func newClientWrapper(config *EnvelopeConfig) (*clientWrapper, error) {
+	client, err := newVaultClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Vault transit path is configurable. "path", "/path", "path/" and "/path/"
+	// are the same.
+	transit := defaultTransitPath
+	if config.TransitPath != "" {
+		transit = config.TransitPath
+	}
+
+	// auth path is configurable. "path", "/path", "path/" and "/path/" are the same.
+	auth := defaultAuthPath
+	if config.AuthPath != "" {
+		auth = config.AuthPath
+	}
+	wrapper := &clientWrapper{
+		client:      client,
+		encryptPath: path.Join("v1", transit, "encrypt"),
+		decryptPath: path.Join("v1", transit, "decrypt"),
+		authPath:    path.Join(auth),
+	}
+
+	// Set token for the vaultapi.client.
+	if len(config.Token) != 0 {
+		client.SetToken(config.Token)
+	} else {
+		err = wrapper.refreshToken(config)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapper, nil
+}
+
+func newVaultClient(config *EnvelopeConfig) (*vaultapi.Client, error) {
+	vaultConfig := vaultapi.DefaultConfig()
+	vaultConfig.Address = config.Address
+
+	tlsConfig := &vaultapi.TLSConfig{
+		CACert:        config.CACert,
+		ClientCert:    config.ClientCert,
+		ClientKey:     config.ClientKey,
+		TLSServerName: config.TLSServerName,
+	}
+	if err := vaultConfig.ConfigureTLS(tlsConfig); err != nil {
+		return nil, err
+	}
+
+	return vaultapi.NewClient(vaultConfig)
+}
+
+// Get token by login and set the value to vaultapi.Client.
+func (c *clientWrapper) refreshToken(config *EnvelopeConfig) error {
+	switch {
+	case config.ClientCert != "" && config.ClientKey != "":
+		token, err := c.tlsToken(config)
+		if err != nil {
+			return err
+		}
+		c.client.SetToken(token)
+	case config.RoleID != "":
+		token, err := c.appRoleToken(config)
+		if err != nil {
+			return err
```

Comment:
Same as above. Please wrap this error with some context.

Reply:
done